### PR TITLE
Support Gemini asset functions 

### PR DESCRIPTION
**Pull request by orta**
A mutation like:

```graphql
mutation {
  requestCredentialsForAssetUpload(input: {name: "threebody-staging", acl: "private", clientMutationId: "1231"}) {
    asset {
      signature
      credentials
      policy_encoded
    }
  }
}
```

Will successfully get you the S3 credentials. 

## TODO

- [x] Get naming right
- [x] Tests
- [x] Docs, for one-offs via dataloader

### CODE REVIEW

**Comment by orta** [comment]
Just gonna do some inline comments now

File: .vscode/launch.json [hunk 1]
```diff
@@ -1,41 +1,16 @@
...
+      "request": "launch",
+      "port": 5858,
+      "address": "localhost",
+      "stopOnEntry": false,
+      "runtimeExecutable": null,
+      "runtimeArgs": ["--inspect-brk", "./node_modules/.bin/jest", "-i"],
+      "cwd": "${workspaceRoot}"
+    }
+  ]
+}
```
Comment (orta) [inline .vscode/launch.json]: No-one is really using these (as we have the debugger launching during dev mode), so I've simplified this

File: .vscode/settings.json [hunk 2]
```diff
@@ -1,13 +1,8 @@
-// Place your settings in this file to overwrite default and user settings.
 {
-  "editor.rulers": [
-      120
-  ],
+  "editor.rulers": [120],
   "editor.tabSize": 2,
-  "javascript.validate.enable": false,
-  "flow.useNPMPackagedFlow": true,
+  "javascript.validate.enable": true,
```
Comment (orta) [inline .vscode/settings.json]: Removes the flow rules, and enables the JavaScript validation - this turns on the inline TS warnings

File: docs/dataloaders.md [hunk 3]
```diff
@@ -0,0 +1,83 @@
+### The 1 paragraph explaination of DataLoader
```
Comment (orta) [inline docs/dataloaders.md]: Quick explanation of DataLoader usage

File: lib/loaders/loaders_without_authentication/gravity.js [hunk 4]
```diff
@@ -0,0 +1,20 @@
+import { gravityLoaderWithoutAuthenticationFactory as gravityLoader } from "../api"
```
Comment (orta) [inline lib/loaders/loaders_without_authentication/gravity.js]: Splits `lib/loaders/loaders_without_authentication/index.js` into three files: gravity, positron and gemini

File: package.json [hunk 5]
```diff
@@ -72,7 +71,9 @@
     "source-map-support": "^0.4.3"
   },
   "devDependencies": {
+    "@types/jest": "^19.2.2",
     "@types/lodash": "^4.14.74",
+    "@types/node": "^8.0.31",
```
Comment (orta) [inline package.json]: Needed now that we're getting inline type reporting

File: package.json [hunk 6]
```diff
@@ -94,7 +95,7 @@
   "jest": {
     "testRegex": "/test/.*.js$",
     "setupTestFrameworkScriptFile": "<rootDir>/test/helper.js",
-    "testPathIgnorePatterns": ["/node_modules/", "/test/helper.js", "/test/utils.js", "/test/__mocks__"]
+    "testPathIgnorePatterns": ["/node_modules/", "/test/helper.js", "/test/utils.js", "/test/gql.js", "/test/__mocks__"]
```
Comment (orta) [inline package.json]: I've added a `gql` function that does nothing. This is enough for prettier and our schema-based vscode extensions, so in an upcoming PR, I will add the ability to set that in the way we do for Emission.

File: schema/me/saved_artworks.js [hunk 7]
```diff
@@ -5,7 +5,7 @@ import { CollectionType, collectionResolver } from "schema/collection"
 export default {
   type: CollectionType,
   resolve: (root, options, request, { fieldNodes, rootValue }) => {
-    const { accessToken, userID } = (rootValue: any)
```
Comment (orta) [inline schema/me/saved_artworks.js]: Removing some unused typing ( due to removal of flow)

File: test/schema/me/consignments/__snapshots__/add_asset_to_submission_mutation.js.snap [hunk 8]
```diff
@@ -1,3 +1,12 @@
 // Jest Snapshot v1, https://goo.gl/fbAQLP
```
Comment (orta) [inline test/schema/me/consignments/__snapshots__/add_asset_to_submission_mutation.js.snap]: The past tests were not set up right

File: package.json (hunk 6 above)
Comment (alloy) [reply package.json]: Nice 👌

File: docs/dataloaders.md (hunk 3 above)
Comment (mzikherman) [reply docs/dataloaders.md]: typo: **explanation**

File: lib/loaders/api/loader_one_off_factory.js [hunk 9]
```diff
@@ -0,0 +1,18 @@
+// @ts-check
+
+import DataLoader from "dataloader"
+import { loaderInterface } from "./loader_interface"
+
+/**
+ * This factory provides a short-cut system for our data loader system, it provides
+ * and uncached result that is close the the equivilent to calling the fetch request directly.
```
Comment (mzikherman) [inline lib/loaders/api/loader_one_off_factory.js]: typo: close **to**

File: schema/asset_uploads/create_asset_request_mutation.js [hunk 10]
```diff
@@ -0,0 +1,49 @@
+// @ts-check
+
+import { GraphQLString, GraphQLNonNull, GraphQLObjectType } from "graphql"
+import { mutationWithClientMutationId } from "graphql-relay"
+
+export const CredentialsType = new GraphQLObjectType({
+  name: "Credentials",
+  description: "An asset which is assigned to a consignment submission",
```
Comment (mzikherman) [inline schema/asset_uploads/create_asset_request_mutation.js]: 🍝

File: lib/loaders/loaders_without_authentication/gravity.js (hunk 4 above)
Comment (mzikherman) [reply lib/loaders/loaders_without_authentication/gravity.js]: 👍

File: lib/loaders/api/loader_one_off_factory.js [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+ *
+ * @param {(string, any) => Promise<any>} api a function that performs an API request
+ * @param {string} path the path for the API call
+ * @param {any} options values which should be passed into the fetch request
+ */
+
+export const loaderOneOffFactory = (api, path, options) => {
```
Comment (mzikherman) [inline lib/loaders/api/loader_one_off_factory.js]: Nice!

File: schema/asset_uploads/create_asset_request_mutation.js [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+  fields: {
+    credentials: {
```
Comment (mzikherman) [inline schema/asset_uploads/create_asset_request_mutation.js]: aw yea cc @izakp Metaphysics getting the S3 key from the newly added `credentials`

File: schema/asset_uploads/create_asset_request_mutation.js [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+      description: "The key to use with S3.",
+      type: new GraphQLNonNull(GraphQLString),
+    },
+    policy_encoded: {
+      description: "The convection submission ID",
```
Comment (mzikherman) [inline schema/asset_uploads/create_asset_request_mutation.js]: 🍝

**Comment by mzikherman** [comment]
Looks great @orta ! 😎 

There's a few typos/copy-pastas in some field descriptions, feel free to merge/fix at your leisure. 👍

File: docs/dataloaders.md [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+
+* The `loader factory` - there are three loader factories: two provide caching, which should be your default and
+  one that provides the same API but for specific one-off problems. They all end up exposing the same API, so your
+  tests should be the same.
+
+  - [`loader_without_authentication_factory`][no_auth_loader] for API requests which can call the `api` directly and be safely cached.  
+  - [`loader_with_authentication_factory`][auth_loader] for API requests that _could_ require a call for an authentication token (like the examples in [Adding a New Microservice][new_micro].)
+  - [`loader_one_off_factory`][one_off_loader] for API requests which have custom auth schemes, or need to ignore cache completely.
+
+* The `loader` themselves. These are a set of functions which [are exposed][loaders] as properties on the root object exposed during the the resolving stages of our graphQL implementation.
```
Comment (alloy) [inline docs/dataloaders.md]: > the ~~the~~ resolving stages

File: .vscode/settings.json (hunk 2 above)
Comment (alloy) [reply .vscode/settings.json]: 😎

File: docs/dataloaders.md (hunk 3 above)
Comment (alloy) [reply docs/dataloaders.md]: Noice 👌

File: schema/asset_uploads/create_asset_request_mutation.js (hunk 12 above)
Comment (alloy) [reply schema/asset_uploads/create_asset_request_mutation.js]: These are one-off keys, meant for just 1 upload, right?

**Comment by alloy** [comment]
The only thing I’d like to have answered is if those returned credentials are indeed 1 off, otherwise just needs a rebase and I’m 👍 on merging this.

File: schema/asset_uploads/create_asset_request_mutation.js (hunk 12 above)
Comment (orta) [reply schema/asset_uploads/create_asset_request_mutation.js]: Yeah, they're size limited too

**Comment by orta** [comment]
I'm gonna merge this when green, I have a waiting to ship PR that adds the other gemini route that I'll do all the typo fixes in - thanks all

**Pull request merged**